### PR TITLE
Support session.range in local testing

### PR DESCRIPTION
**Pull request by sfc-gh-stan**
Please answer these questions before submitting your pull requests. Thanks!

1. What GitHub issue is this PR addressing? Make sure that there is an accompanying issue to your PR.

   Fixes #NNNN

2. Fill out the following pre-review checklist:

   - [ ] I am adding a new automated test(s) to verify correctness of my new code
   - [ ] I am adding new logging messages
   - [ ] I am adding a new telemetry message
   - [ ] I am adding new credentials
   - [ ] I am adding a new dependency

3. Please describe how your code solves the related issue.

   Add support for session range in local testing to unblock integ testing for join, filter, etc.

### CODE REVIEW

File: src/snowflake/snowpark/mock/snowflake_data_type.py [hunk 1]
```diff
@@ -95,7 +95,10 @@ def __getitem__(self, item):
 
     def __setitem__(self, key, value):
         super().__setitem__(key, value)
-        self.sf_types[key] = value.sf_type
+        if isinstance(
```
Comment (sfc-gh-stan) [inline src/snowflake/snowpark/mock/snowflake_data_type.py:98]: Are all values here expected to be ColumnEmulator? cc: @sfc-gh-yixie 
I ran into when this is just `pd.Series`.

File: tests/integ/test_dataframe.py [hunk 2]
```diff
@@ -898,6 +903,9 @@ def test_filter(session):
     expected = []
     assert res == expected
 
+
+def test_filter_with_sql_str(session):
```
Comment (sfc-gh-stan) [inline tests/integ/test_dataframe.py:907]: I separated the part from the original `test_filter` that specifies condition as `str` into a this new test.

File: src/snowflake/snowpark/mock/mock_analyzer.py [hunk 3]
```diff
@@ -546,15 +545,6 @@ def do_resolve_with_resolved_children(
     ) -> MockExecutionPlan:
         if isinstance(logical_plan, MockExecutionPlan):
             return logical_plan
-
-        # Re-wrap SelectStatement into a MockSelectStatement
```
Comment (sfc-gh-stan) [inline src/snowflake/snowpark/mock/mock_analyzer.py:550]: We shouldn't have this. Let's completely separate `MockSelectStatment` and `SelectStatement`, these two should not be used together.

File: tests/integ/test_dataframe.py (hunk 2 above)
Comment (sfc-gh-aling) [reply tests/integ/test_dataframe.py:907]: out of curiosity, a string condition will be passed into the execution plan as it is?

I'm maintaining a list of unsupported cases so thanks for catching this

File: src/snowflake/snowpark/session.py [hunk 4]
```diff
@@ -1684,8 +1683,10 @@ def range(self, start: int, end: Optional[int] = None, step: int = 1) -> DataFra
         if self.sql_simplifier_enabled:
             df = DataFrame(
                 self,
-                SelectStatement(
-                    from_=SelectSnowflakePlan(range_plan, analyzer=self._analyzer),
+                self._analyzer.create_SelectStatement(
+                    from_=self._analyzer.create_SelectSnowflakePlan(
+                        range_plan, analyzer=self._analyzer
+                    ),
```
Comment (sfc-gh-aling) [inline src/snowflake/snowpark/session.py:1689]: is this trying to unify mock and non-mock?

what's difference between calling `SelectStatement` and calling `self._analyzer.create_SelectStatement` or non-mocking execution, will there be any side-effect?

(and same question to `create_SelectSnowflakePlan`)

File: src/snowflake/snowpark/mock/snowflake_data_type.py (hunk 1 above)
Comment (sfc-gh-aling) [reply src/snowflake/snowpark/mock/snowflake_data_type.py:98]: I think so, what's the case that you run into `pd.Series`?

File: src/snowflake/snowpark/mock/mock_analyzer.py [hunk 5]
```diff
@@ -141,7 +141,6 @@
     MockSelectable,
     MockSelectableEntity,
     MockSelectExecutionPlan,
-    MockSelectSnowflakePlan,
```
Comment (sfc-gh-aling) [inline src/snowflake/snowpark/mock/mock_analyzer.py:144]: based upon our discussion, this class can be removed from `mock_select_statement.py`?

Comment (sfc-gh-stan) [reply src/snowflake/snowpark/mock/mock_analyzer.py:144]: Yes, I could include this in this PR.

File: src/snowflake/snowpark/session.py (hunk 4 above)
Comment (sfc-gh-stan) [reply src/snowflake/snowpark/session.py:1689]: The difference lies in `self._analyzer`, `MockAnalyzer` should only create `MockSelectable` and real `Analyzer` should only create real `Selectable`.
Yes this is for unifying codepath and minimizing change, although I don't have a strong preference here.

Comment (sfc-gh-aling) [reply src/snowflake/snowpark/session.py:1689]: I checked the real `Analyzer`, this change should be good!

**Review by sfc-gh-aling — APPROVED** [review]

**Review by sfc-gh-aling — APPROVED** [review]

**Pull request merged**